engine/execution/ingestion/block_queue/queue.go

```diff
@@ -0,0 +1,382 @@
+	if executable.IsComplete() {
+		// executables might contain other siblings, but won't contain "executable",
+		// which is the block itself, that's because executables are created
+		// from OnBlockExecuted(
+		executables = []*entity.ExecutableBlock{executable}
+	}
+
+	return missingCollections, executables, nil
+}
+
+// OnCollection is called when a new collection is received
+// It returns a list of executable blocks that contains the collection
+func (q *BlockQueue) OnCollection(collection *flow.Collection) ([]*entity.ExecutableBlock, error) {
+	q.Lock()
+	defer q.Unlock()
+	// when a collection is received, we find the blocks the collection is included in,
+	// and check if the blocks become executable.
+	// Note a collection could be included in multiple blocks, so receiving a collection
+	// might trigger multiple blocks to be executable.
+
+	// check if the collection is for any block in the queue
+	colID := collection.ID()
+	colInfo, ok := q.collections[colID]
+	if !ok {
+		// no block in the queue includes this collection
+		return nil, nil
+	}
+
+	if colInfo.Collection.IsCompleted() {
+		// the collection is already received, no action needed because an action must
+		// have been returned when the collection is first received.
+		return nil, nil
+	}
+
+	// update collection
+	colInfo.Collection.Transactions = collection.Transactions
+
+	// check if any block, which includes this collection, become executable
+	executables := make([]*entity.ExecutableBlock, 0, len(colInfo.IncludedIn))
+	for _, block := range colInfo.IncludedIn {
+		if !block.IsComplete() {
+			continue
+		}
+		executables = append(executables, block)
+	}
+
+	if len(executables) == 0 {
+		return nil, nil
+	}
+
+	return executables, nil
+}
+
+// OnBlockExecuted is called when a block is executed
+// It returns a list of executable blocks (usually its child blocks)
+// The caller has to ensure OnBlockExecuted is not called in a wrong order, such as
+// OnBlockExecuted(childBlock) being called before OnBlockExecuted(parentBlock).
+func (q *BlockQueue) OnBlockExecuted(
+	blockID flow.Identifier,
+	commit flow.StateCommitment,
+) ([]*entity.ExecutableBlock, error) {
+	q.Lock()
+	defer q.Unlock()
+
+	return q.onBlockExecuted(blockID, commit)
+}
+
+func (q *BlockQueue) onBlockExecuted(
+	blockID flow.Identifier,
+	commit flow.StateCommitment,
+) ([]*entity.ExecutableBlock, error) {
+	// when a block is executed, the child block might become executable
+	// we also remove it from all the indexes
+
+	// remove block
+	block, ok := q.blocks[blockID]
+	if !ok {
+		return nil, nil
+	}
+
+	// sanity check
+	// if a block exists in the queue and is executed, then its parent block
+	// must not exist in the queue, otherwise the state is inconsistent
+	_, parentExists := q.blocks[block.Block.Header.ParentID]
+	if parentExists {
+		return nil, fmt.Errorf("parent block %s of block %s is in the queue",
+			block.Block.Header.ParentID, blockID)
```

Comment:
Sanity checks like this make me think that this should use the `irrecoverable.SignalerContext` to throw this error and shutdown the component entirely.

Reply:
We could leave the decision to the caller. The caller can decide how to handle this exception, either throw as a irrecoverable error or handle it differently.